src/cart/crt_hg_proc.c

```diff
@@ -412,9 +435,22 @@ crt_hg_unpack_header(hg_handle_t handle, struct crt_rpc_priv *rpc_priv,
 		D_ERROR("crt_proc_common_hdr failed rc: %d.\n", rc);
 		D_GOTO(out, rc);
 	}
-	/* Clients never decode requests. */
+
+	/* Sync the HLC. Clients never decode requests. */
 	D_ASSERT(crt_is_service());
-	(void)crt_hlc_get_msg(rpc_priv->crp_req_hdr.cch_hlc);
+	rc = crt_hlc_get_msg(rpc_priv->crp_req_hdr.cch_hlc, NULL /* hlc_out */);
+	if (rc != 0) {
+		REPORT_HLC_SYNC_ERR("failed to sync HLC for request: opc=%x ts="
+				    DF_U64" from=%u\n",
+				    rpc_priv->crp_req_hdr.cch_opc,
+				    rpc_priv->crp_req_hdr.cch_hlc,
+				    rpc_priv->crp_req_hdr.cch_src_rank);
+		/* Fail all but SWIM requests. */
+		if (!crt_opc_is_swim(rpc_priv->crp_req_hdr.cch_opc))
+			D_GOTO(out, rc);
+		rc = 0;
```

Comment:
hmm, it will fail all RPC except swim RPC if send_ts (hlc in msg) - local ts > crt_hlc_epsilon (which is 1 second by default)?
I am not sure if it is practical in real system?

Reply:
> hmm, it will fail all RPC except swim RPC if send_ts (hlc in msg) - local ts > crt_hlc_epsilon (which is 1 second by default)?

Yes.

> I am not sure if it is practical in real system?

It can't be guaranteed. But AFAIK, 1 s is quite practical for real systems with proper clock synchronization.

Reply:
i'm afraid if enable this patch by default, some daos user may find it does not work in their environment/config?

what's the main purpose of add the requirement in this patch? (is it must needed?)
Or could it (failing of RPC) be optional? for example can force the ts check and rejection only when a certain ENV is set?

Reply:
Unfortunately, without this patch, if a system's clocks violate epsilon, workloads may fail without obvious clues pointing to the clocks. It may be difficult to figure out why they fail. Also, without this patch, if a faulty client sends a very high HLC timestamp to the system, it may push the system's clocks way ahead into the future. This large epsilon violation may cause HLC logical part overflows, causing errors, delays, or silent corruptions, depending on how we will handle these overflows.

Reply:
OK, i see, thanks. probably fine to land it first.